File: packages/code-storage-go/commit.go
package storage

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/google/uuid"
)

const maxChunkBytes = 4 * 1024 * 1024

type commitOperation struct {
	Path      string
	ContentID string
	Mode      GitFileMode
	Operation string
	Source    io.Reader
}

func (b *CommitBuilder) normalize() error {
	b.options.TargetBranch = strings.TrimSpace(b.options.TargetBranch)
	b.options.TargetRef = strings.TrimSpace(b.options.TargetRef)
	b.options.CommitMessage = strings.TrimSpace(b.options.CommitMessage)

	if b.options.TargetBranch != "" {
		branch, err := normalizeBranchName(b.options.TargetBranch)
		if err != nil {
			return err
		}
		b.options.TargetBranch = branch
	} else if b.options.TargetRef != "" {
		branch, err := normalizeLegacyTargetRef(b.options.TargetRef)
		if err != nil {
			return err
		}
		b.options.TargetBranch = branch
	} else {
		return errors.New("createCommit targetBranch is required")
	}

	if b.options.CommitMessage == "" {
		return errors.New("createCommit commitMessage is required")
	}

	if strings.TrimSpace(b.options.Author.Name) == "" || strings.TrimSpace(b.options.Author.Email) == "" {
		return errors.New("createCommit author name and email are required")
	}
	b.options.Author.Name = strings.TrimSpace(b.options.Author.Name)
	b.options.Author.Email = strings.TrimSpace(b.options.Author.Email)

	b.options.ExpectedTargetSHA = preferredString(b.options.ExpectedTargetSHA, b.options.ExpectedHeadSHA)
	b.options.TargetIsEphemeral = preferredBool(b.options.TargetIsEphemeral, b.options.Ephemeral)
	b.options.BaseIsEphemeral = preferredBool(b.options.BaseIsEphemeral, b.options.EphemeralBase)
	b.options.BaseBranch = strings.TrimSpace(b.options.BaseBranch)
	if b.options.BaseBranch != "" && strings.HasPrefix(b.options.BaseBranch, "refs/") {
		return errors.New("createCommit baseBranch must not include refs/ prefix")
	}

	if b.options.BaseIsEphemeral != nil && *b.options.BaseIsEphemeral && b.options.BaseBranch == "" {
		return errors.New("createCommit baseIsEphemeral requires baseBranch")
	}

	if b.options.Committer != nil {
		if strings.TrimSpace(b.options.Committer.Name) == "" || strings.TrimSpace(b.options.Committer.Email) == "" {
			return errors.New("createCommit committer name and email are required when provided")
		}
		b.options.Committer.Name = strings.TrimSpace(b.options.Committer.Name)
		b.options.Committer.Email = strings.TrimSpace(b.options.Committer.Email)
	}

	return nil
}

// AddFile adds a file to the commit.
func (b *CommitBuilder) AddFile(path string, source io.Reader, options *CommitFileOptions) *CommitBuilder {
	if b.err != nil {
		return b
	}
	if err := b.ensureNotSent(); err != nil {
		b.err = err
		return b
	}
	normalizedPath, err := normalizePath(path)
	if err != nil {
		b.err = err
		return b
	}
	if source == nil {
		b.err = errors.New("unsupported content source; expected binary data")
		return b
	}

	mode := GitFileModeRegular
	if options != nil && options.Mode != "" {
		mode = options.Mode
	}

	b.ops = append(b.ops, commitOperation{
		Path:      normalizedPath,
		ContentID: uuid.NewString(),
		Mode:      mode,
		Operation: "upsert",
		Source:    source,
	})
	return b
}

// AddFileFromBytes adds a binary file.
func (b *CommitBuilder) AddFileFromBytes(path string, contents []byte, options *CommitFileOptions) *CommitBuilder {
	if b.err != nil {
		return b
	}
	return b.AddFile(path, bytes.NewReader(contents), options)
}

// AddFileFromString adds a text file.
func (b *CommitBuilder) AddFileFromString(path string, contents string, options *CommitTextFileOptions) *CommitBuilder {
	if b.err != nil {
		return b
	}
	encoding := "utf-8"
	if options != nil && options.Encoding != "" {
		encoding = options.Encoding
	}
	encoding = strings.ToLower(strings.TrimSpace(encoding))
	if encoding != "utf8" && encoding != "utf-8" {
		b.err = errors.New("unsupported encoding: " + encoding)
		return b
	}
	if options == nil {
		return b.AddFile(path, strings.NewReader(contents), nil)
	}
	return b.AddFile(path, strings.NewReader(contents), &options.CommitFileOptions)
}

// DeletePath removes a file or directory.
func (b *CommitBuilder) DeletePath(path string) *CommitBuilder {
	if b.err != nil {
		return b
	}
	if err := b.ensureNotSent(); err != nil {
		b.err = err
		return b
	}
	normalizedPath, err := normalizePath(path)
	if err != nil {
		b.err = err
		return b
	}
	b.ops = append(b.ops, commitOperation{
		Path:      normalizedPath,
		ContentID: uuid.NewString(),
		Operation: "delete",
	})
	return b
}

// Err returns any error accumulated during builder operations.
func (b *CommitBuilder) Err() error {
	return b.err
}

// Send finalizes the commit.
func (b *CommitBuilder) Send(ctx context.Context) (CommitResult, error) {
	if b.err != nil {
		return CommitResult{}, b.err
	}
	if err := b.ensureNotSent(); err != nil {
		return CommitResult{}, err
	}
	b.sent = true

	if strings.TrimSpace(b.repoID) == "" {
		return CommitResult{}, errors.New("createCommit repository id is required")
	}
	if b.client == nil {
		return CommitResult{}, errors.New("createCommit client is required")
	}

	ttl := resolveCommitTTL(b.options.InvocationOptions, defaultTokenTTL)
	jwtToken, err := b.client.generateJWT(b.repoID, RemoteURLOptions{Permissions: []Permission{PermissionGitWrite}, TTL: ttl, RefPolicies: b.options.RefPolicies})
	if err != nil {
		return CommitResult{}, err
	}

	metadata := buildCommitMetadata(b.options, b.ops)

	pipeReader, pipeWriter := io.Pipe()
	encoder := json.NewEncoder(pipeWriter)
	encoder.SetEscapeHTML(false)

	go func() {
		defer pipeWriter.Close()
		if err := encoder.Encode(metadataEnvelope{Metadata: metadata}); err != nil {
			_ = pipeWriter.CloseWithError(err)
			return
		}

		for _, op := range b.ops {
			if op.Operation != "upsert" {
				continue
			}
			if err := writeBlobChunks(encoder, op.ContentID, op.Source); err != nil {
				_ = pipeWriter.CloseWithError(err)
				return
			}
		}
	}()

	requestURL := b.client.api.basePath() + "/repos/" + url.PathEscape(b.repoID) + "/commit-pack"
	resp, err := doStreamingRequest(ctx, b.client.api.httpClient, http.MethodPost, requestURL, jwtToken, pipeReader)
	if err != nil {
		return CommitResult{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		fallback := "createCommit request failed (" + itoa(resp.StatusCode) + " " + resp.Status + ")"
		statusMessage, statusLabel, refUpdate, err := parseCommitPackError(resp, fallback)
		if err != nil {
			return CommitResult{}, err
		}
		return CommitResult{}, newRefUpdateError(statusMessage, statusLabel, refUpdate)
	}

	var ack commitPackAck
	if err := decodeJSON(resp, &ack); err != nil {
		return CommitResult{}, err
	}

	return buildCommitResult(ack)
}

func (b *CommitBuilder) ensureNotSent() error {
	if b.sent {
		return errors.New("createCommit builder cannot be reused after send")
	}
	return nil
}

func buildCommitMetadata(options CommitOptions, ops []commitOperation) *commitMetadataPayload {
	files := make([]fileEntryPayload, 0, len(ops))
	for _, op := range ops {
		entry := fileEntryPayload{
			Path:      op.Path,
			ContentID: op.ContentID,
			Operation: op.Operation,
		}
		if op.Operation == "upsert" && op.Mode != "" {
			entry.Mode = string(op.Mode)
		}
		files = append(files, entry)
	}

	metadata := &commitMetadataPayload{
		TargetBranch:  options.TargetBranch,
		CommitMessage: options.CommitMessage,
		Author: authorInfo{
			Name:  options.Author.Name,
			Email: options.Author.Email,
		},
		Files: files,
	}

	if options.ExpectedTargetSHA != "" {
		metadata.ExpectedTargetSHA = options.ExpectedTargetSHA
	}
	if options.BaseBranch != "" {
		metadata.BaseBranch = options.BaseBranch
	}
	if options.Committer != nil {
		metadata.Committer = &authorInfo{
			Name:  options.Committer.Name,
			Email: options.Committer.Email,
		}
	}
	metadata.TargetIsEphemeral = options.TargetIsEphemeral
	metadata.BaseIsEphemeral = options.BaseIsEphemeral

	return metadata
}

func writeBlobChunks(encoder *json.Encoder, contentID string, reader io.Reader) error {
	buf := make([]byte, maxChunkBytes)
	var pending []byte
	for {
		n, err := reader.Read(buf)
		if n > 0 {
			if pending != nil {
				payload := blobChunkEnvelope{
					BlobChunk: blobChunkPayload{
						ContentID: contentID,
						Data:      base64.StdEncoding.EncodeToString(pending),
						EOF:       false,
					},
				}
				if err := encoder.Encode(payload); err != nil {
					return err
				}
			}
			pending = append(pending[:0], buf[:n]...)
		}
		if err == io.EOF {
			if pending == nil {
				payload := blobChunkEnvelope{
					BlobChunk: blobChunkPayload{
						ContentID: contentID,
						Data:      "",
						EOF:       true,
					},
				}
				return encoder.Encode(payload)
			}
			payload := blobChunkEnvelope{
				BlobChunk: blobChunkPayload{
					ContentID: contentID,
					Data:      base64.StdEncoding.EncodeToString(pending),
					EOF:       true,
				},
			}
			return encoder.Encode(payload)
		}
		if err != nil {
			return err
		}
	}
}

func normalizePath(path string) (string, error) {
	path = strings.TrimSpace(path)
	if path == "" {
		return "", errors.New("file path must be a non-empty string")
	}
	return strings.TrimPrefix(path, "/"), nil
}

func normalizeBranchName(value string) (string, error) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return "", errors.New("createCommit targetBranch is required")
	}
	if strings.HasPrefix(trimmed, "refs/heads/") {
		branch := strings.TrimSpace(strings.TrimPrefix(trimmed, "refs/heads/"))
		if branch == "" {
			return "", errors.New("createCommit targetBranch is required")
		}
		return branch, nil
	}
	if strings.HasPrefix(trimmed, "refs/") {
		return "", errors.New("createCommit targetBranch must not include refs/ prefix")
	}
	return trimmed, nil
}

func normalizeLegacyTargetRef(ref string) (string, error) {
	trimmed := strings.TrimSpace(ref)
	if trimmed == "" {
		return "", errors.New("createCommit targetRef is required")
	}
	if !strings.HasPrefix(trimmed, "refs/heads/") {
		return "", errors.New("createCommit targetRef must start with refs/heads/")
	}
	branch := strings.TrimSpace(strings.TrimPrefix(trimmed, "refs/heads/"))
	if branch == "" {
		return "", errors.New("createCommit targetRef must include a branch name")
	}
	return branch, nil
}

func resolveCommitTTL(options InvocationOptions, defaultValue time.Duration) time.Duration {
	if options.TTL > 0 {
		return options.TTL
	}
	return defaultValue
}

func doStreamingRequest(ctx context.Context, client *http.Client, method string, url string, jwtToken string, body io.Reader) (*http.Response, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	if client == nil {
		client = http.DefaultClient
	}

	req, err := http.NewRequestWithContext(ctx, method, url, body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+jwtToken)
	req.Header.Set("Content-Type", "application/x-ndjson")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Code-Storage-Agent", userAgent())

	return client.Do(req)
}

File: packages/code-storage-typescript/src/commit.ts
import { buildCommitResult, parseCommitPackError } from './commit-pack';
import { RefUpdateError } from './errors';
import type { CommitPackAckRaw } from './schemas';
import { commitPackAckSchema } from './schemas';
import {
  type ChunkSegment,
  base64Encode,
  chunkify,
  requiresDuplex,
  toAsyncIterable,
  toRequestBody,
} from './stream-utils';
import type {
  CommitBuilder,
  CommitFileOptions,
  CommitFileSource,
  CommitResult,
  CommitSignature,
  CommitTextFileOptions,
  CreateCommitOptions,
  CreateCommitRefOptions,
} from './types';
import { getUserAgent } from './version';

const DEFAULT_TTL_SECONDS = 60 * 60;
const HEADS_REF_PREFIX = 'refs/heads/';

type NodeBuffer = Uint8Array & { toString(encoding?: string): string };
interface NodeBufferConstructor {
  from(data: Uint8Array): NodeBuffer;
  from(data: string, encoding?: string): NodeBuffer;
  isBuffer(value: unknown): value is NodeBuffer;
}

const BufferCtor: NodeBufferConstructor | undefined = (
  globalThis as { Buffer?: NodeBufferConstructor }
).Buffer;

interface CommitMetadataPayload {
  target_branch: string;
  expected_target_sha?: string;
  base_branch?: string;
  commit_message: string;
  target_is_ephemeral?: boolean;
  base_is_ephemeral?: boolean;
  author: {
    name: string;
    email: string;
  };
  committer?: {
    name: string;
    email: string;
  };
  files: Array<{
    path: string;
    content_id: string;
    operation?: 'upsert' | 'delete';
    mode?: string;
  }>;
}

interface CommitTransportRequest {
  authorization: string;
  signal?: AbortSignal;
  metadata: CommitMetadataPayload;
  blobs: Array<{ contentId: string; chunks: AsyncIterable<ChunkSegment> }>;
}

interface CommitTransport {
  send(request: CommitTransportRequest): Promise<CommitPackAckRaw>;
}

type NormalizedCommitOptions = {
  targetBranch: string;
  commitMessage: string;
  expectedTargetSha?: string;
  baseBranch?: string;
  targetIsEphemeral?: boolean;
  baseIsEphemeral?: boolean;
  author: CommitSignature;
  committer?: CommitSignature;
  signal?: AbortSignal;
  ttl?: number;
};

interface CommitBuilderDeps {
  options: CreateCommitOptions;
  getAuthToken: () => Promise<string>;
  transport: CommitTransport;
}

type FileOperationState = {
  path: string;
  contentId: string;
  mode?: string;
  operation: 'upsert' | 'delete';
  streamFactory?: () => AsyncIterable<Uint8Array>;
};

export class CommitBuilderImpl implements CommitBuilder {
  private readonly options: NormalizedCommitOptions;
  private readonly getAuthToken: () => Promise<string>;
  private readonly transport: CommitTransport;
  private readonly operations: FileOperationState[] = [];
  private sent = false;

  constructor(deps: CommitBuilderDeps) {
    this.options = normalizeCommitOptions(deps.options);
    this.getAuthToken = deps.getAuthToken;
    this.transport = deps.transport;

    const trimmedMessage = this.options.commitMessage?.trim();
    const trimmedAuthorName = this.options.author?.name?.trim();
    const trimmedAuthorEmail = this.options.author?.email?.trim();

    if (!trimmedMessage) {
      throw new Error('createCommit commitMessage is required');
    }
    if (!trimmedAuthorName || !trimmedAuthorEmail) {
      throw new Error('createCommit author name and email are required');
    }
    this.options.commitMessage = trimmedMessage;
    this.options.author = {
      name: trimmedAuthorName,
      email: trimmedAuthorEmail,
    };
    if (typeof this.options.expectedTargetSha === 'string') {
      this.options.expectedTargetSha = this.options.expectedTargetSha.trim();
    }
    if (typeof this.options.baseBranch === 'string') {
      const trimmedBase = this.options.baseBranch.trim();
      if (trimmedBase === '') {
        delete this.options.baseBranch;
      } else {
        if (trimmedBase.startsWith('refs/')) {
          throw new Error(
            'createCommit baseBranch must not include refs/ prefix'
          );
        }
        this.options.baseBranch = trimmedBase;
      }
    }

    if (this.options.baseIsEphemeral && !this.options.baseBranch) {
      throw new Error('createCommit baseIsEphemeral requires baseBranch');
    }
  }

  addFile(
    path: string,
    source: CommitFileSource,
    options?: CommitFileOptions
  ): CommitBuilder {
    this.ensureNotSent();
    const normalizedPath = this.normalizePath(path);
    const contentId = randomContentId();
    const mode = options?.mode ?? '100644';

    this.operations.push({
      path: normalizedPath,
      contentId,
      mode,
      operation: 'upsert',
      streamFactory: () => toAsyncIterable(source),
    });

    return this;
  }

  addFileFromString(
    path: string,
    contents: string,
    options?: CommitTextFileOptions
  ): CommitBuilder {
    const encoding = options?.encoding ?? 'utf8';
    const normalizedEncoding = encoding === 'utf-8' ? 'utf8' : encoding;
    let data: Uint8Array;
    if (normalizedEncoding === 'utf8') {
      data = new TextEncoder().encode(contents);
    } else if (BufferCtor) {
      data = BufferCtor.from(
        contents,
        normalizedEncoding as Parameters<NodeBufferConstructor['from']>[1]
      );
    } else {
      throw new Error(
        `Unsupported encoding "${encoding}" in this environment. Non-UTF encodings require Node.js Buffer support.`
      );
    }
    return this.addFile(path, data, options);
  }

  deletePath(path: string): CommitBuilder {
    this.ensureNotSent();
    const normalizedPath = this.normalizePath(path);
    this.operations.push({
      path: normalizedPath,
      contentId: randomContentId(),
      operation: 'delete',
    });
    return this;
  }

  async send(): Promise<CommitResult> {
    this.ensureNotSent();
    this.sent = true;

    const metadata = this.buildMetadata();
    const blobEntries = this.operations
      .filter((op) => op.operation === 'upsert' && op.streamFactory)
      .map((op) => ({
        contentId: op.contentId,
        chunks: chunkify(op.streamFactory!()),
      }));

    const authorization = await this.getAuthToken();
    const ack = await this.transport.send({
      authorization,
      signal: this.options.signal,
      metadata,
      blobs: blobEntries,
    });
    return buildCommitResult(ack);
  }

  private buildMetadata(): CommitMetadataPayload {
    const files = this.operations.map((op) => {
      const entry: CommitMetadataPayload['files'][number] = {
        path: op.path,
        content_id: op.contentId,
        operation: op.operation,
      };
      if (op.mode) {
        entry.mode = op.mode;
      }
      return entry;
    });

    const metadata: CommitMetadataPayload = {
      target_branch: this.options.targetBranch,
      commit_message: this.options.commitMessage,
      author: {
        name: this.options.author.name,
        email: this.options.author.email,
      },
      files,
    };

    if (this.options.expectedTargetSha) {
      metadata.expected_target_sha = this.options.expectedTargetSha;
    }
    if (this.options.baseBranch) {
      metadata.base_branch = this.options.baseBranch;
    }
    if (this.options.committer) {
      metadata.committer = {
        name: this.options.committer.name,
        email: this.options.committer.email,
      };
    }

    if (typeof this.options.targetIsEphemeral === 'boolean') {
      metadata.target_is_ephemeral = this.options.targetIsEphemeral;
    }
    if (typeof this.options.baseIsEphemeral === 'boolean') {
      metadata.base_is_ephemeral = this.options.baseIsEphemeral;
    }

    return metadata;
  }

  private ensureNotSent(): void {
    if (this.sent) {
      throw new Error('createCommit builder cannot be reused after send()');
    }
  }

  private normalizePath(path: string): string {
    if (!path || typeof path !== 'string' || path.trim() === '') {
      throw new Error('File path must be a non-empty string');
    }
    return path.replace(/^\//, '');
  }
}

export class FetchCommitTransport implements CommitTransport {
  private readonly url: string;

  constructor(config: { baseUrl: string; version: number }) {
    const trimmedBase = config.baseUrl.replace(/\/+$/, '');
    this.url = `${trimmedBase}/api/v${config.version}/repos/commit-pack`;
  }

  async send(request: CommitTransportRequest): Promise<CommitPackAckRaw> {
    const bodyIterable = buildMessageIterable(request.metadata, request.blobs);
    const body = toRequestBody(bodyIterable);

    const init: RequestInit = {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${request.authorization}`,
        'Content-Type': 'application/x-ndjson',
        Accept: 'application/json',
        'Code-Storage-Agent': getUserAgent(),
      },
      body: body as any,
      signal: request.signal,
    };

    if (requiresDuplex(body)) {
      (init as RequestInit & { duplex: 'half' }).duplex = 'half';
    }

    const response = await fetch(this.url, init);

    if (!response.ok) {
      const fallbackMessage = `createCommit request failed (${response.status} ${response.statusText})`;
      const { statusMessage, statusLabel, refUpdate } =
        await parseCommitPackError(response, fallbackMessage);
      throw new RefUpdateError(statusMessage, {
        status: statusLabel,
        message: statusMessage,
        refUpdate,
      });
    }

    const ack = commitPackAckSchema.parse(await response.json());
    return ack;
  }
}

function buildMessageIterable(
  metadata: CommitMetadataPayload,
  blobs: Array<{ contentId: string; chunks: AsyncIterable<ChunkSegment> }>
): AsyncIterable<Uint8Array> {
  const encoder = new TextEncoder();
  return {
    async *[Symbol.asyncIterator]() {
      yield encoder.encode(`${JSON.stringify({ metadata })}\n`);
      for (const blob of blobs) {
        for await (const segment of blob.chunks) {
          const payload = {
            blob_chunk: {
              content_id: blob.contentId,
              data: base64Encode(segment.chunk),
              eof: segment.eof,
            },
          };
          yield encoder.encode(`${JSON.stringify(payload)}\n`);
        }
      }
    },
  };
}

function randomContentId(): string {
  const cryptoObj = globalThis.crypto;
  if (cryptoObj && typeof cryptoObj.randomUUID === 'function') {
    return cryptoObj.randomUUID();
  }
  const random = Math.random().toString(36).slice(2);
  return `cid-${Date.now().toString(36)}-${random}`;
}

function normalizeCommitOptions(
  options: CreateCommitOptions
): NormalizedCommitOptions {
  return {
    targetBranch: resolveTargetBranch(options),
    commitMessage: options.commitMessage,
    expectedTargetSha: options.expectedTargetSha ?? options.expectedHeadSha,
    baseBranch: options.baseBranch,
    targetIsEphemeral: resolveOptionalBoolean(
      options.targetIsEphemeral,
      options.ephemeral
    ),
    baseIsEphemeral: resolveOptionalBoolean(
      options.baseIsEphemeral,
      options.ephemeralBase
    ),
    author: options.author,
    committer: options.committer,
    signal: options.signal,
    ttl: options.ttl,
  };
}

function resolveOptionalBoolean(
  preferred: boolean | undefined,
  deprecated: boolean | undefined
): boolean | undefined {
  if (typeof preferred === 'boolean') return preferred;
  if (typeof deprecated === 'boolean') return deprecated;
  return undefined;
}

function resolveTargetBranch(options: CreateCommitOptions): string {
  const branchCandidate =
    typeof options.targetBranch === 'string' ? options.targetBranch.trim() : '';
  if (branchCandidate) {
    return normalizeBranchName(branchCandidate);
  }
  if (hasTargetRef(options)) {
    return normalizeTargetRef(options.targetRef);
  }
  throw new Error('createCommit targetBranch is required');
}

function normalizeBranchName(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error('createCommit targetBranch is required');
  }
  if (trimmed.startsWith(HEADS_REF_PREFIX)) {
    const branch = trimmed.slice(HEADS_REF_PREFIX.length).trim();
    if (!branch) {
      throw new Error('createCommit targetBranch is required');
    }
    return branch;
  }
  if (trimmed.startsWith('refs/')) {
    throw new Error('createCommit targetBranch must not include refs/ prefix');
  }
  return trimmed;
}

function normalizeTargetRef(ref: string): string {
  const trimmed = ref.trim();
  if (!trimmed) {
    throw new Error('createCommit targetRef is required');
  }
  if (!trimmed.startsWith(HEADS_REF_PREFIX)) {
    throw new Error('createCommit targetRef must start with refs/heads/');
  }
  const branch = trimmed.slice(HEADS_REF_PREFIX.length).trim();
  if (!branch) {
    throw new Error('createCommit targetRef must include a branch name');
  }
  return branch;
}

function hasTargetRef(
  options: CreateCommitOptions
): options is CreateCommitRefOptions {
  return typeof (options as CreateCommitRefOptions).targetRef === 'string';
}

export function createCommitBuilder(deps: CommitBuilderDeps): CommitBuilder {
  return new CommitBuilderImpl(deps);
}

export function resolveCommitTtlSeconds(options?: { ttl?: number }): number {
  if (typeof options?.ttl === 'number' && options.ttl > 0) {
    return options.ttl;
  }
  return DEFAULT_TTL_SECONDS;
}
